src/object/srv_obj_migrate.c

```diff
@@ -1766,6 +1860,129 @@ migrate_one_destroy(struct migrate_one *mrone)
 	D_FREE(mrone);
 }
 
+enum {
+	OBJ_ULT = 1,
+	DKEY_ULT = 2,
+};
+
+/* Check if there are enough resource for the migration to proceed. */
+static int
+migrate_system_enter(struct migrate_pool_tls *tls, int tgt_idx)
+{
+	uint32_t tgt_cnt = 0;
+	int	 rc = 0;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+	D_ASSERTF(tgt_idx < dss_tgt_nr, "tgt idx %d tgt nr %u\n", tgt_idx, dss_tgt_nr);
+
+	tgt_cnt = atomic_load(&tls->mpt_obj_ult_cnts[tgt_idx]) +
+		  atomic_load(&tls->mpt_dkey_ult_cnts[tgt_idx]);
+
+	while ((tls->mpt_inflight_max_ult / dss_tgt_nr) <= tgt_cnt) {
+		D_DEBUG(DB_REBUILD, "tgt%d:%u max %u\n",
+			tgt_idx, tgt_cnt, tls->mpt_inflight_max_ult / dss_tgt_nr);
+		ABT_mutex_lock(tls->mpt_inflight_mutex);
+		ABT_cond_wait(tls->mpt_inflight_cond, tls->mpt_inflight_mutex);
+		ABT_mutex_unlock(tls->mpt_inflight_mutex);
+		if (tls->mpt_fini)
+			D_GOTO(out, rc = -DER_SHUTDOWN);
+
+		tgt_cnt = atomic_load(&tls->mpt_obj_ult_cnts[tgt_idx]) +
+			  atomic_load(&tls->mpt_dkey_ult_cnts[tgt_idx]);
+	}
+
+	atomic_fetch_add(&tls->mpt_obj_ult_cnts[tgt_idx], 1);
+out:
+	return rc;
+}
+
+static int
+migrate_tgt_enter(struct migrate_pool_tls *tls)
+{
+	uint32_t dkey_cnt = 0;
+	int	 rc = 0;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id != 0);
+
+	dkey_cnt = atomic_load(tls->mpt_tgt_dkey_ult_cnt);
+	while (tls->mpt_inflight_max_ult / 2 <= dkey_cnt) {
+		D_DEBUG(DB_REBUILD, "tgt %u max %u\n", dkey_cnt, tls->mpt_inflight_max_ult);
+
+		ABT_mutex_lock(tls->mpt_inflight_mutex);
+		ABT_cond_wait(tls->mpt_inflight_cond, tls->mpt_inflight_mutex);
```

Comment:
just confirm in the test did it really trigger such cond_wait and wakeup?

Reply:
yes, it does. all these adjustments and fixes came from my local rebuilding 1M objects test.